src/server/pkg/storage/tracker/tracker.go

```diff
@@ -0,0 +1,150 @@
+package tracker
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/client/pkg/require"
+	"github.com/pachyderm/pachyderm/src/server/pkg/dbutil"
+)
+
+var (
+	ErrObjectExists  = errors.Errorf("object exists")
+	ErrDanglingRef   = errors.Errorf("the operation would create a dangling reference")
+	ErrTombstone     = errors.Errorf("cannot create object because it is marked as a tombstone")
+	ErrNotTombstone  = errors.Errorf("object cannot be deleted because it is not marked as a tombstone")
+	ErrSelfReference = errors.Errorf("object cannot reference itself")
+)
+
+// Tracker tracks objects and their references to one another.
+type Tracker interface {
+	// CreateObject creates an object with id=id, and pointers to everything in pointsTo
+	// It errors with ErrObjectExists if the object already exists.  Callers may be able to ignore this.
+	// It errors with ErrDanglingRef if any of the elements in pointsTo do not exist
+	// It errors with ErrTombstone if the object exists and is marked as a tombstone. Callers should retry until
+	// they successfully create the object.
+	// CreateObject should always be called *before* adding an object to an auxillary store.
+	CreateObject(ctx context.Context, id string, pointsTo []string, ttl time.Duration) error
+
+	// SetTTLPrefix sets the expiration time to current_time + ttl for all objects with ids starting with prefix
+	SetTTLPrefix(ctx context.Context, prefix string, ttl time.Duration) (time.Time, error)
+	// TODO: thoughts on these?
+	// SetTTL(ctx context.Context, id string, ttl time.Duration) error
+	// SetTTLBatch(ctx context.Context, ids []string, ttl time.Duration) error
+
+	// GetDownstream gets all objects immediately downstream of (pointed to by) object with id
+	GetDownstream(ctx context.Context, id string) ([]string, error)
+
+	// GetUpstream gets all objects immediately upstream of (pointing to) the object with id
+	GetUpstream(ctx context.Context, id string) ([]string, error)
+
+	// MarkTombstone causes the object to appear deleted.  It cannot be created, and objects referencing it cannot
+	// be created until it is deleted.
+	// It errors if id has other objects referencing it.
+	// Marking something as a tombstone which is already a tombstone is not an error
+	MarkTombstone(ctx context.Context, id string) error
+
+	// FinishDelete deletes the object
+	// It is an error to call DeleteObject without calling MarkTombstone.
+	FinishDelete(ctx context.Context, id string) error
+
+	// IterateExpired calls cb with all the objects which have expired or are tombstones.
+	IterateExpired(ctx context.Context, cb func(id string) error) error
```

Comment:
This should probably be renamed. It is iterating over the unreferenced objects that have expired, not all of the expired objects. This should probably be something that points more to being eligible for deletion, so maybe `IterateDeletes`?